[FMV][GlobalOpt] Statically resolve calls to versioned functions.

labrinea opened the pull request:
To deduce whether the optimization is legal we need to compare the target
features between caller and callee versions. The criteria for bypassing
the resolver are the following:

 * If the callee's feature set is a subset of the caller's feature set,
   then the callee is a candidate for direct call.

 * Among such candidates the one of highest priority is the best match
   and it shall be picked, unless there is a version of the callee with
   higher priority than the best match which cannot be picked from a
   higher priority caller (directly or through the resolver).

 * For every higher priority callee version than the best match, there
   is a higher priority caller version whose feature set availability
   is implied by the callee's feature set.

labrinea committed:
[FMV][GlobalOpt] Bypass the IFunc Resolver of MultiVersioned functions.

To deduce whether the optimization is legal we need to compare the target
features between caller and callee versions. The criteria for bypassing
the resolver are the following:

 * If the callee's feature set is a subset of the caller's feature set,
   then the callee is a candidate for direct call.

 * Among such candidates the one of highest priority is the best match
   and it shall be picked, unless there is a version of the callee with
   higher priority than the best match which cannot be picked from a
   higher priority caller (directly or through the resolver).

 * For every higher priority callee version than the best match, there
   is a higher priority caller version whose feature set availability
   is implied by the callee's feature set.

Example:

Callers and Callees are ordered in decreasing priority.
The arrows indicate successful call redirections.

  Caller        Callee      Explanation
=========================================================================
mops+sve2 --+--> mops       all the callee versions are subsets of the
            |               caller but mops has the highest priority
            |
     mops --+    sve2       between mops and default callees, mops wins

      sve        sve        between sve and default callees, sve wins
                            but sve2 does not have a high priority caller

  default -----> default    sve (callee) implies sve (caller),
                            sve2(callee) implies sve (caller),
                            mops(callee) implies mops(caller)

jroelofs commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2462,6 +2462,228 @@ DeleteDeadIFuncs(Module &M,
+        return constFoldInst(&BB->front(), Root, Seed, Br->getParent());
+      }
+    } else {
+      // Don't know how to handle. Bail.
```
think it's worth having a stat counter for the number of not-yet-handled resolvers observed?

labrinea replied:
There can be various reasons of early exit, not only unhandled instruction in the resolver. For example see line 2613 and 2676. Do we need to differentiate them? Line 2613 suggests the resolver didn't look as expected, line 2676 suggests we failed to constant fold, which basically implies the resolver didn't look as expected (the handled instruction operands weren't constants), so as this line here. I don't see value in collecting stats for each of these cases treating it a separate instance.

jroelofs replied:
ok

andrewcarlotti commented:
The testcase and description look valid, although the description seems a little more conservative than necessary.

If you were to add a `bar._Mmops` version, would it's call be optimised to `test_multi_bb_resolver._Mmops`? And would that also enable the call from `bar.default` to be optimised to point to `test_multi_bb_resolver.default`?

labrinea commented:
> The testcase and description look valid, although the description seems a little more conservative than necessary.

"Subset" is perhaps not the right terminology. I meant to say the feature set is "less or equal to". In other words "implied".
 
> If you were to add a `bar._Mmops` version, would it's call be optimised to `test_multi_bb_resolver._Mmops`? 

Yes

> And would that also enable the call from `bar.default` to be optimised to point to `test_multi_bb_resolver.default`?

No, because there is no corresponding caller to pick `test_multi_bb_resolver._Msve2`. The feature set of `test_multi_bb_resolver._Msve2` is not implied by the feature set of `bar._sve`.

labrinea commented:
@jroelofs, aside from semantics which I'll clarify with Andrew, I am thinking that the constant folding part might be redundant and whether `FeatureBitset` is an alternative to the TargetParser.

labrinea commented:
I've removed the constant folding since it was indeed unnecessary computation. Also I found a testcase which exposes a potential issue: a function version (for example AES) having the same target features as the default. In TargetParser we have `ExtensionInfo->DependentFeatures` to model this information (I believe @ilinpv introduced this change) and it seems that for some entries we don't add the corresponding backend feature (perhaps intentionally?). I will try to address this in a separate ticket.

labrinea commented:
> I've removed the constant folding since it was indeed unnecessary computation. Also I found a testcase which exposes a potential issue: a function version (for example AES) having the same target features as the default. In TargetParser we have `ExtensionInfo->DependentFeatures` to model this information (I believe @ilinpv introduced this change) and it seems that for some entries we don't add the corresponding backend feature (perhaps intentionally?). I will try to address this in a separate ticket.

@andrewcarlotti already has a patch for ACLE https://github.com/ARM-software/acle/pull/315 which removes `pmull` since it is implied by `aes`, therefore we can add the missing `+aes` in ExtensionInfo->DependentFeatures for `aes`.

labrinea commented:
Hi, @jroelofs. This patch has been quiet in a while and the reason is I have been collaborating with @tmatheson-arm in refactoring the TargetParser as a preliminary step. However the landscape doesn't seem to be changing much, it's mostly NFC. What I mean is that for this patch to work we need to be able to express every FMV feature in the LLVM IR attribute "target-features". This isn't the case for a few FMV features and I believe it won't be easy to change. For example the FMV feature AES in ExtensionInfo has no DependentFeatures because the SubtargetFeature for AES is fused with PMULL, meaning we can't link them (the fmv feature and the subtarget counterpart) together as they mean different things. I do not want to rely on the cleanup in https://github.com/ARM-software/acle/pull/315 as features may diverge again between FMV and the AAch64 compiler backend. Spliting up backend features will break backwards compatibility in LLVM IR (because they will also have to be renamed), so maybe we need some other way to propagate FMV information from clang to LLVM (not via the target-features attribute). Is this worth an RFC, like introducing a new attribute for FMV in IR? Is this optimization still valuable to you? Shall we pursue this further or just give up?

jroelofs commented:
Yes, this optimization is still interesting to me, but I’m not in a rush. We can take our time on it. I have a non-FMV project I need to focus on over the summer, but I expect to pick up work on things in this area in the fall.

labrinea commented:
Folks, I will be rebasing this patch soon. I am thinking that the HasFMV target hooks may not be suitable. What if we are targeting a platform where https://github.com/llvm/llvm-project/pull/87942 would yield `-target-feature +fmv` and this optimization modifies calls to ifuncs that are not FMV ifuncs? Do we need perhaps to introduce a new call attribute?

jroelofs commented:
Why do we need to avoid optimizing non-fmv ifuncs?

labrinea commented:
> Why do we need to avoid optimizing non-fmv ifuncs?

This optimization is only valid for FMV resolvers because we know what they do: they check the presence of FMV feature bits set by the runtime. Other ifunc resolvers may do completely different things that have nothing to do with the target-features of the caller/callee.

labrinea committed:
Merge branch 'main' into fmv-resolve-ifunc

labrinea commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.h:
```diff
@@ -88,6 +88,10 @@ class AArch64TTIImpl : public BasicTTIImplBase<AArch64TTIImpl> {
   unsigned getInlineCallPenalty(const Function *F, const CallBase &Call,
                                 unsigned DefaultCallPenalty) const;
 
+  bool hasFMV() const { return ST->hasFMV(); }
```
Oops, leftover. To be removed

labrinea committed:
Changes from last revision:

* clang format
* remove leftover target hook hasFMV after rebase
* remove filter in regression test after rebase

labrinea commented on llvm/test/Transforms/GlobalOpt/resolve-fmv-ifunc.ll:
```diff
@@ -0,0 +1,271 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter "call i32 @(test_single_bb_resolver|test_multi_bb_resolver|test_caller_feats_not_implied|test_same_priority_callers)" --version 4
```
test_same_priority_callers is another leftover to be removed

labrinea committed:
Merge branch 'main' into fmv-resolve-ifunc

labrinea committed:
Changes from last revision:

Use FMV priority mask when sorting candidates

labrinea committed:
Changes from last revision

Allow the optimization when the caller is non FMV but the attributes match.

labrinea committed:
Changes from last revision

Add a problematic test case

labrinea committed:
clang-format

labrinea commented on llvm/include/llvm/TargetParser/AArch64FeatPriorities.inc:
```diff
@@ -0,0 +1,66 @@
+#define AARCH64_FEAT_PRIORITIES_INC_H
+
+// Function Multi Versioning feature priorities.
+enum FeatPriorities {
```
@jroelofs I will raise a PR to address the ACLE changes https://github.com/ARM-software/acle/pull/370. I am thinking this enum is perhaps worth separating from this patch and I am considering moving it there. My question is do we need to preserve backwards ABI compatibility, or shall I reorder the entries of the existing enum in llvm/TargetParser/AArch64CPUFeatures.inc (which also happens to have annoying gaps due to removed FMV features)?

jroelofs replied:
The ABI break from re-ordering/re-numbering these doesn't bother me until FMV moves out of beta.

labrinea replied:
Okay, thanks for lettimg me know. I'll ask what others think.

labrinea replied:
I am thinking if we add new FMV features whose priority sits between existing ones then it would displace them. So maybe having this additional enum is inevitable.

labrinea replied:
Raised https://github.com/llvm/llvm-project/pull/121921, resolving this thread.

goldsteinn commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2642,6 +2642,145 @@ DeleteDeadIFuncs(Module &M,
+  } else if (auto *Phi = dyn_cast<PHINode>(V)) {
+    for (unsigned I = 0, E = Phi->getNumIncomingValues(); I != E; ++I)
+      collectVersions(Phi->getIncomingValue(I), Versions);
+  }
```
If you fail to walk use-def chain and find the actual function, could it not be an issue in that we might be ignorant to the actual best match?

labrinea replied:
If I understood your question correctly, line 2700 accounts for that:
```
if (Callees.empty())
      continue;
```

goldsteinn replied:
Err no I mean if there is a return value but you can't collect the assosiated function, couldn't we optimize to a static call when a better match existed? I.e im proposing adding:
```
  ...
  } else if (auto *Phi = dyn_cast<PHINode>(V)) {
    for (unsigned I = 0, E = Phi->getNumIncomingValues(); I != E; ++I)
      if(!collectVersions(Phi->getIncomingValue(I), Versions)) return false;
    return true;
  }
  else
    return false;
```

then `if(!collectVersions(...)) { // skip transform entirely }`

labrinea replied:
Ah I see now, to abort earlier. Sounds right, I could do that.

labrinea committed:
Merge branch 'main' into fmv-resolve-ifunc

labrinea committed:
Minor adjustments in test file:
* Rename caller function names
* Remove whitespace between function version declarations.

labrinea committed:
Add a test for parsing fmv features whose backend feature has alternative name.

labrinea committed:
Early exit if collectVersions fails to walk the use-def chain at any point.

labrinea commented:
FYI I am ready to rebase this on https://github.com/llvm/llvm-project/pull/122192. I have addressed @goldsteinn 's comment locally, made some adjustments and added a few more tests.

labrinea committed:
Only consider the highest priority callee when the caller is non-FMV.
Add a corresponding test case.

labrinea committed:
Merge branch 'main' into fmv-resolve-ifunc

labrinea commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2641,6 +2641,165 @@ DeleteDeadIFuncs(Module &M,
+      Changed = true;
+    }
+    if (IF.use_empty() ||
+        all_of(IF.users(), [](User *U) { return isa<GlobalAlias>(U); }))
```
That is probably a leftover from the time we had ifunc aliases. Subject to removal.

labrinea commented on llvm/test/Transforms/GlobalOpt/resolve-fmv-ifunc.ll:
```diff
@@ -0,0 +1,365 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter "call i32 @(test_single_bb_resolver|test_multi_bb_resolver|test_caller_feats_not_implied|test_non_fmv_caller|test_priority|test_alternative_names)" --version 4
+; RUN: opt --passes=globalopt -o - -S < %s | FileCheck %s
```
Idea: perhaps it would be useful to have an integration test in clang at -O3 to show that this optimization is working as expected on C/C++ source?

jroelofs commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1850,6 +1850,13 @@ class TargetTransformInfo {
 
+  /// Returns a bitmask constructed from the target-features or fmv-features
+  /// metadata of a function.
+  uint64_t getFeatureMask(Function &F) const;
```
can it be `const Function &`? likewise for `isMultiversionedFunction`

jroelofs reviewed (approved):
LGTM

labrinea committed:
Merge branch 'main' into fmv-resolve-ifunc

labrinea committed:
Make the argument to getFeatureMask and isMultiversionedFunction a constant reference.

jplehr commented:
Hi, there are two more bots red:
https://lab.llvm.org/buildbot/#/builders/140/builds/14935
https://lab.llvm.org/staging/#/builders/130/builds/12489

Let me know if you need help from my side.

labrinea commented:
Please review they fix here https://github.com/llvm/llvm-project/pull/123322

labrinea commented:
The llvm test suite builder https://lab.llvm.org/buildbot/#/builders/199/builds/1002 is broken too and I am investigating if it relates to this change.

labrinea commented:
Verdict for the llvm-test-suite regression: What I am thinking is that the compiler assumes that we have `predres` because of `mcpu=neoverse-v2`, but in reality we don't have it on the target, and since the optimization is skipping the runtime check, the test fails. Perhaps it's only safe to perform the optimization when the caller is a versioned function. In this example the caller is main, which is a regular function.

labrinea commented:
Yeah, maybe. How shall we change the test though?

labrinea commented:
https://lab.llvm.org/buildbot/#/builders/198/builds/1232/steps/5/logs/stdio

> cmake -G Ninja ../llvm/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=True '-DLLVM_LIT_ARGS='"'"'-v'"'"'' -DCMAKE_INSTALL_PREFIX=../stage1.install '-DCMAKE_C_FLAGS='"'"'-mcpu=neoverse-v2'"'"'' '-DCMAKE_CXX_FLAGS='"'"'-mcpu=neoverse-v2'"'"'' -DLLVM_ENABLE_LLD=True -DMLIR_INCLUDE_INTEGRATION_TESTS=True -DMLIR_RUN_ARM_SVE_TESTS=True '-DLLVM_ENABLE_PROJECTS=clang-tools-extra;lld;mlir;llvm;clang;flang' -DLLVM_ENABLE_RUNTIMES=compiler-rt

jroelofs commented:
Looks like that builder is:
```
Fujitsu FX700 48x A64FX; RAM 185GB OS: Ubuntu 22.04.5 LTS Kernel: 6.8.0-1019-aws #21-Ubuntu SMP Wed Nov 6 21:53:02 UTC 2024 Compiler: clang version 18.1.8 (http://git.linaro.org/toolchain/jenkins-scripts.git e5def089cd9f5aa71524f82fef301ca66eaa38d2) System Linker: GNU ld (GNU Binutils for Ubuntu) 2.38 C Library: ldd (Ubuntu GLIBC 2.35-0ubuntu3.8) 2.35 CMake: cmake version 3.28.6
```

https://lab.llvm.org/buildbot/#/workers/158

so `-mcpu=a64fx` would be much more appropriate. @rengolin do you know who runs the linaro bots?

labrinea commented:
Since it's late Friday and the buildbots are still red, shall we temporarily land https://github.com/llvm/llvm-project/pull/123383 and revisit this early next week?

jroelofs commented:
SGTM

labrinea commented:
Update: @DavidSpickett has fixed the host info on the linaro-g4 builders and it is now correctly showing Neoverse-V2. From my end I tried the following test on Graviton4:
```
$ cat specres.c 
#include <stdio.h>

__attribute__((target_version("predres"))) void f(void) { printf("Has FEAT_SPECRES\n"); }
__attribute__((target_version("default"))) void f(void) { printf("Does not have FEAT_SPECRES\n"); }

int main() {
    f();
    asm volatile (
      "cfp rctx, x0" "\n"
      "dvp rctx, x1" "\n"
      "cpp rctx, x2" "\n"
    );
    return 0;
}
$ ./release/bin/clang --target=aarch64-linux-gnu -mcpu=neoverse-v2 specres.c -O0 -rtlib=compiler-rt
$ ./a.out 
Does not have FEAT_SPECRES
Illegal instruction (core dumped)
```
The runtime detection suggests `predres` isn't available. I am now trying to understand whether this has to do with reading ID_AA64ISAR1_EL1 from EL0 and whether the CFP/DVP/CPP RCTX , Xt instructions are valid at EL0. ArmARM says yes, but

> In AArch64 state, EL0 access to the System instructions is controlled by:
• When HCR_EL2.{E2H, TGE} is not {1, 1}, SCTLR_EL1.EnRCTX.
• When HCR_EL2.{E2H, TGE} == {1, 1}, SCTLR_EL2.EnRCTX.

labrinea commented:
Reading more on https://www.kernel.org/doc/html/latest/arch/arm64/cpu-feature-registers.html my understanding is that some feature bits from ID_AA64ISAR1_EL1 may be visible, but specres doesn't seem to be one of them. Also my naive interpretation of https://github.com/torvalds/linux/blob/master/arch/arm64/kernel/cpufeature.c#L236 is that SPECRES is hidden.

labrinea commented:
> Accessing a restricted system register from an application generates an exception and ends up in SIGILL being delivered to the process. The infrastructure hooks into the exception handler and emulates the operation if the source belongs to the supported system register space.

@jroelofs I have confirmed the above with the linux kernel folks. SPECRES and LS64 are hidden features so reading those with MRS always yields zero. This shouldn't affect your platform though.

jroelofs commented:
So should we disable/XFAIL or mark that test UNSUPPORTED on Linux platforms?

labrinea commented:
I am thinking that the SPECRES instructions would trap when executed from EL0 on Darwin too, just like [this example](https://github.com/llvm/llvm-project/pull/87939#issuecomment-2603203462). We just don't have a bot to catch it.

My reading of ArmARM makes me think that LS64 is also unusable from EL0, but I may be wrong:
> SCTLR_EL1.EnALS, bit [56] : When FEAT_LS64 is implemented:
>   When the Effective value of HCR_EL2.{E2H, TGE} is not {1, 1}, traps execution of an LD64B or ST64B instruction at EL0 to EL1.
>
> SCTLR_EL1.EnAS0, bit [55] : When FEAT_LS64_ACCDATA is implemented:
>   When the Effective value of HCR_EL2.{E2H, TGE} is not {1, 1}, traps execution of an ST64BV0 instruction at EL0 to EL1.
>
> SCTLR_EL1.EnASR, bit [54] : When FEAT_LS64_V is implemented:
>   When the Effective value of HCR_EL2.{E2H, TGE} is not {1, 1}, traps execution of an ST64BV instruction at EL0 to EL1.
 
 The pseudocode in CheckLDST64BEnabled(), CheckST64BV0Enabled() and CheckST64BVEnabled() suggests the same.
 
 So perhaps we should remove those features. I am trying to clarify internally.
